# Challenge 2

In this challenge we will continue working with the `Pokemon` dataset. We will attempt solving a slightly more complex problem in which we will practice the iterative data analysis process you leaned in [this video](https://www.youtube.com/watch?v=xOomNicqbkk).

The problem statement is as follows:

**You are at a Pokemon black market planning to buy a Pokemon for battle. All Pokemon are sold at the same price and you can only afford to buy one. You cannot choose which specific Pokemon to buy. However, you can specify the type of the Pokemon - one type that exists in either `Type 1` or `Type 2`. Which type should you choose in order to maximize your chance of receiving a good Pokemon?**

To remind you about the 3 steps of iterative data analysis, they are:

1. Setting Expectations
1. Collecting Information
1. Reacting to Data / Revising Expectations

Following the iterative process, we'll guide you in completing the challenge.


## Problem Solving Iteration 1

In this iteration we'll analyze the problem and identify the breakthrough. The original question statement is kind of vague because we don't know what a *good pokemon* really means as represented in the data. We'll start by understanding the dataset and see if we can find some insights.

In [1]:
# Import libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import OneHotEncoder

In [2]:
# Importing the dataset
pokemon = pd.read_csv('pokemon.csv')
pokemon

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 13 columns]

From the data it seems whether a pokemon is good depends on its abilities as represented in the fields of `HP`, `Attack`, `Defense`, `Sp. Atk`, `Sp. Def`, `Speed`, and `Total`. We are not sure about `Generation` and `Legendary` because they are not necessarily the decisive factors of the pokemon abilities.

But `HP`, `Attack`, `Defense`, `Sp. Atk`, `Sp. Def`, `Speed`, and `Total` are a lot of fields! If we look at them all at once it's very complicated. This isn't Mission Impossible but it's ideal that we tackle this kind of problem after we learn Machine Learning (which you will do in Module 3). For now, is there a way to consolidate the fields we need to look into?

Fortunately there seems to be a way. It appears the `Total` field is computed based on the other 6 fields. But we need to prove our theory. If we can approve there is a formula to compute `Total` based on the other 6 abilities, we only need to look into `Total`.

We have the following expectation now:

#### The `Total` field is computed based on `HP`, `Attack`, `Defense`, `Sp. Atk`, `Sp. Def`, and `Speed`.

We need to collect the following information:

* **What is the formula to compute `Total`?**
* **Does the formula work for all pokemon?**

In the cell below, make a hypothesis on how `Total` is computed and test your hypothesis.

In [3]:
# your code here
sum_stats = pokemon[['HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed']].sum(axis=1)
(pokemon['Total'] == sum_stats).all()

#https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.all.html#pandas.DataFrame.all

True

**based on my research: https://www.kaggle.com/datasets/abcsds/pokemon , total is the sum of all others **



## Problem Solving Iteration 2

Now that we have consolidated the abilities fields, we can update the problem statement. The new problem statement is:

### Which pokemon type is most likely to have the highest `Total` value?

In the updated problem statement, we assume there is a certain relationship between the `Total` and the pokemon type. But we have two *type* fields (`Type 1` and `Type 2`) that have string values. In data analysis, string fields have to be transformed to numerical format in order to be analyzed. 

In addition, keep in mind that `Type 1` always has a value but `Type 2` is sometimes empty (having the `NaN` value). Also, the pokemon type we choose may be either in `Type 1` or `Type 2`.

Now our expectation is:

#### `Type 1` and `Type 2` string variables need to be converted to numerical variables in order to identify the relationship between `Total` and the pokemon type.

The information we need to collect is:

#### How to convert two string variables to numerical?

Let's address the first question first. You can use a method called **One Hot Encoding** which is frequently used in machine learning to encode categorical string variables to numerical. The idea is to gather all the possible string values in a categorical field and create a numerical field for each unique string value. Each of those numerical fields uses `1` and `0` to indicate whether the data record has the corresponding categorical value. A detailed explanation of One Hot Encoding can be found in [this article](https://hackernoon.com/what-is-one-hot-encoding-why-and-when-do-you-have-to-use-it-e3c6186d008f). You will formally learn it in Module 3.

For instance, if a pokemon has `Type 1` as `Poison` and `Type 2` as `Fire`, then its `Poison` and `Fire` fields are `1` whereas all other fields are `0`. If a pokemon has `Type 1` as `Water` and `Type 2` as `NaN`, then its `Water` field is `1` whereas all other fields are `0`.

#### In the next cell, use One Hot Encoding to encode `Type 1` and `Type 2`. Use the pokemon type values as the names of the numerical fields you create.

The new numerical variables you create should look like below:

![One Hot Encoding](../images/one-hot-encoding.png)

In [4]:
# your code here
# Type 1
unique_types_t1 = pokemon['Type 1'].unique().reshape(-1, 1)
encoder_t1 = OneHotEncoder().fit(unique_types_t1)
encoded_t1 = encoder_t1.transform(pokemon[['Type 1']]).toarray()

# Type 2
unique_types_t2 = pokemon['Type 2'].unique().reshape(-1, 1)
encoder_t2 = OneHotEncoder().fit(unique_types_t2)
encoded_t2 = encoder_t2.transform(pokemon[['Type 2']]).toarray()

# get column names, adjust to remove the initalc column part 
feature_names_t1 = encoder_t1.get_feature_names_out(['Type 1'])
feature_names_t2 = encoder_t2.get_feature_names_out(['Type 2'])
feature_names = np.concatenate((feature_names_t1, feature_names_t2))
feature_names = [name.split('_')[1] for name in feature_names]
#https://stackoverflow.com/questions/54570947/feature-names-from-onehotencoder

# Combine into a single df
encoded_concatenated = np.concatenate((encoded_t1, encoded_t2), axis=1)
onehot_encoded = pd.DataFrame(encoded_concatenated, columns=feature_names)

# remove last column (NaN, group by to remove duplicate columns)
#https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html
onehot_encoded = onehot_encoded.iloc[:, :-1].groupby(level=0, axis=1).sum().astype(int)

# rearrange columns to match the png provided
onehot_encoded = onehot_encoded.reindex(columns=['Grass', 'Poison', 'Fire', 'Flying', 'Dragon', 'Water', 'Bug', 'Normal', 'Electric', 'Ground', 'Fairy', 'Fighting','Psychic', 'Rock', 'Steel', 'Ice', 'Ghost', 'Dark'])

# Display the encoded types
onehot_encoded.head(5)

/opt/anaconda3/lib/python3.11/site-packages/sklearn/base.py:432: UserWarning: X has feature names, but OneHotEncoder was fitted without feature names
  warnings.warn(
/opt/anaconda3/lib/python3.11/site-packages/sklearn/base.py:432: UserWarning: X has feature names, but OneHotEncoder was fitted without feature names
  warnings.warn(
/var/folders/pr/f4jl4tln3yn9fd_1yjrwwkp40000gn/T/ipykernel_27463/1499198518.py:25: FutureWarning: DataFrame.groupby with axis=1 is deprecated. Do `frame.T.groupby(...)` without axis instead.
  onehot_encoded = onehot_encoded.iloc[:, :-1].groupby(level=0, axis=1).sum().astype(int)


Grass  Poison  Fire  Flying  Dragon  Water  Bug  Normal  Electric  Ground  \
0      1       1     0       0       0      0    0       0         0       0   
1      1       1     0       0       0      0    0       0         0       0   
2      1       1     0       0       0      0    0       0         0       0   
3      1       1     0       0       0      0    0       0         0       0   
4      0       0     1       0       0      0    0       0         0       0   

   Fairy  Fighting  Psychic  Rock  Steel  Ice  Ghost  Dark  
0      0         0        0     0      0    0      0     0  
1      0         0        0     0      0    0      0     0  
2      0         0        0     0      0    0      0     0  
3      0         0        0     0      0    0      0     0  
4      0         0        0     0      0    0      0     0

In [5]:
type_columns = pokemon[['Type 1', 'Type 2']]
type_encoder = OneHotEncoder(drop=None).fit(type_columns)

column_names = type_encoder.get_feature_names_out(input_features=type_columns.columns)
column_names = [name.split('_')[1] for name in column_names]

type_encoded = pd.DataFrame(type_encoder.transform(type_columns).toarray(), columns=column_names)
type_encoded.head()

type_encoded = type_encoded.iloc[:, :-1].T.groupby(level=0).sum().astype(int)

# rearrange columns to match the png provided
type_encoded = type_encoded.reindex(columns=['Grass', 'Poison', 'Fire', 'Flying', 'Dragon', 'Water', 'Bug', 'Normal', 'Electric', 'Ground', 'Fairy', 'Fighting','Psychic', 'Rock', 'Steel', 'Ice', 'Ghost', 'Dark'])

# Display the encoded types
type_encoded.head(5)

Grass  Poison  Fire  Flying  Dragon  Water  Bug  Normal  Electric  \
Bug         NaN     NaN   NaN     NaN     NaN    NaN  NaN     NaN       NaN   
Dark        NaN     NaN   NaN     NaN     NaN    NaN  NaN     NaN       NaN   
Dragon      NaN     NaN   NaN     NaN     NaN    NaN  NaN     NaN       NaN   
Electric    NaN     NaN   NaN     NaN     NaN    NaN  NaN     NaN       NaN   
Fairy       NaN     NaN   NaN     NaN     NaN    NaN  NaN     NaN       NaN   

          Ground  Fairy  Fighting  Psychic  Rock  Steel  Ice  Ghost  Dark  
Bug          NaN    NaN       NaN      NaN   NaN    NaN  NaN    NaN   NaN  
Dark         NaN    NaN       NaN      NaN   NaN    NaN  NaN    NaN   NaN  
Dragon       NaN    NaN       NaN      NaN   NaN    NaN  NaN    NaN   NaN  
Electric     NaN    NaN       NaN      NaN   NaN    NaN  NaN    NaN   NaN  
Fairy        NaN    NaN       NaN      NaN   NaN    NaN  NaN    NaN   NaN

## Problem Solving Iteration 3

Now we have encoded the pokemon types, we will identify the relationship between `Total` and the encoded fields. Our expectation is:

#### There are relationships between `Total` and the encoded pokemon type variables and we need to identify the correlations.

The information we need to collect is:

#### How to identify the relationship between `Total` and the encoded pokemon type fields?

There are multiple ways to answer this question. The easiest way is to use correlation. In the cell below, calculate the correlation of `Total` to each of the encoded fields. Rank the correlations and identify the #1 pokemon type that is most likely to have the highest `Total`.

In [6]:
# your code here
pokemon_encoded = pd.concat([pokemon['Total'], onehot_encoded], axis=1)
correlation_matrix = pokemon_encoded.corr()
total_correlation = correlation_matrix['Total'].sort_values(ascending=False)
display(total_correlation)

Total       1.000000
Dragon      0.229705
Psychic     0.124688
Steel       0.109703
Fire        0.078726
Fighting    0.077786
Ice         0.060248
Flying      0.059383
Dark        0.056154
Rock        0.032731
Electric    0.020971
Ground      0.015060
Ghost       0.003641
Water      -0.021665
Fairy      -0.036698
Grass      -0.052592
Poison     -0.090441
Normal     -0.105331
Bug        -0.145781
Name: Total, dtype: float64

In [7]:
total_correlation.abs().sort_values(ascending=False)

Total       1.000000
Dragon      0.229705
Bug         0.145781
Psychic     0.124688
Steel       0.109703
Normal      0.105331
Poison      0.090441
Fire        0.078726
Fighting    0.077786
Ice         0.060248
Flying      0.059383
Dark        0.056154
Grass       0.052592
Fairy       0.036698
Rock        0.032731
Water       0.021665
Electric    0.020971
Ground      0.015060
Ghost       0.003641
Name: Total, dtype: float64

> **DRAGON**

# Bonus Question

Say now you can choose both `Type 1` and `Type 2` of the pokemon. In order to receive the best pokemon, which types will you choose?

In [8]:
#Identifying all the combinations of Type2 associated to Type1 Dragon:
dragon = pokemon.loc[pokemon['Type 1'] == 'Dragon']
dragon_t2 = dragon['Type 2'].unique()
dragon_t2

array([nan, 'Flying', 'Fairy', 'Psychic', 'Ground', 'Fire', 'Electric',
       'Ice'], dtype=object)

In [9]:
for type_2_value in dragon_t2:
    print(type_2_value)
    dragon_pokemon_with_type_2 = pokemon[(pokemon['Type 1'] == 'Dragon') & (pokemon['Type 2'] == type_2_value)]
    display(dragon_pokemon_with_type_2)


nan


Empty DataFrame
Columns: [#, Name, Type 1, Type 2, Total, HP, Attack, Defense, Sp. Atk, Sp. Def, Speed, Generation, Legendary]
Index: []

Flying


#                     Name  Type 1  Type 2  Total   HP  Attack  \
161  149                Dragonite  Dragon  Flying    600   91     134   
365  334                  Altaria  Dragon  Flying    490   75      70   
408  373                Salamence  Dragon  Flying    600   95     135   
409  373  SalamenceMega Salamence  Dragon  Flying    700   95     145   
425  384                 Rayquaza  Dragon  Flying    680  105     150   
426  384    RayquazaMega Rayquaza  Dragon  Flying    780  105     180   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
161       95      100      100     80           1      False  
365       90       70      105     80           3      False  
408       80      110       80    100           3      False  
409      130      120       90    120           3      False  
425       90      150       90     95           3       True  
426      100      180      100    115           3       True

Fairy


#                 Name  Type 1 Type 2  Total  HP  Attack  Defense  \
366  334  AltariaMega Altaria  Dragon  Fairy    590  75     110      110   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
366      110      105     80           3      False

Psychic


#               Name  Type 1   Type 2  Total  HP  Attack  Defense  \
417  380             Latias  Dragon  Psychic    600  80      80       90   
418  380  LatiasMega Latias  Dragon  Psychic    700  80     100      120   
419  381             Latios  Dragon  Psychic    600  80      90       80   
420  381  LatiosMega Latios  Dragon  Psychic    700  80     130      100   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
417      110      130    110           3       True  
418      140      150    110           3       True  
419      130      110    110           3       True  
420      160      120    110           3       True

Ground


#                   Name  Type 1  Type 2  Total   HP  Attack  Defense  \
491  443                  Gible  Dragon  Ground    300   58      70       45   
492  444                 Gabite  Dragon  Ground    410   68      90       65   
493  445               Garchomp  Dragon  Ground    600  108     130       95   
494  445  GarchompMega Garchomp  Dragon  Ground    700  108     170      115   
794  718       Zygarde50% Forme  Dragon  Ground    600  108     100      121   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
491       40       45     42           4      False  
492       50       55     82           4      False  
493       80       85    102           4      False  
494      120       95     92           4      False  
794       81       95     95           6       True

Fire


#      Name  Type 1 Type 2  Total   HP  Attack  Defense  Sp. Atk  \
706  643  Reshiram  Dragon   Fire    680  100     120      100      150   

     Sp. Def  Speed  Generation  Legendary  
706      120     90           5       True

Electric


#    Name  Type 1    Type 2  Total   HP  Attack  Defense  Sp. Atk  \
707  644  Zekrom  Dragon  Electric    680  100     150      120      120   

     Sp. Def  Speed  Generation  Legendary  
707      100     90           5       True

Ice


#                Name  Type 1 Type 2  Total   HP  Attack  Defense  \
710  646              Kyurem  Dragon    Ice    660  125     130       90   
711  646  KyuremBlack Kyurem  Dragon    Ice    700  125     170      100   
712  646  KyuremWhite Kyurem  Dragon    Ice    700  125     120       90   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
710      130       90     95           5       True  
711      120       90     95           5       True  
712      170      100     95           5       True

> **we would choose Dragon and Electric or Fire, because there is only one card availabale at 680, best worst case, playing it safe**